### PR TITLE
Annotations and masks workflows

### PR DESCRIPTION
**Pull request by kheffah**
# Part 1: Annotations to masks workflow

**Overview:** 
This includes tools to parse annotations from an item (slide) into masks to use in training and evaluating imaging algorithms. The basic functionality involves mapping annotation rectangles (potentially rotated) and polygons to regions-of-interest (ROIs) and making one mask per region of interest. This extends on some of the workflows described in Amgad et al, 2019:

_Mohamed Amgad, Habiba Elfandy, Hagar Hussein, ..., Jonathan Beezley, Deepak R Chittajallu, David Manthey, David A Gutman, Lee A D Cooper, Structured crowdsourcing enables convolutional segmentation of histology images, Bioinformatics, , btz083, https://doi.org/10.1093/bioinformatics/btz083_

This slide used as a test example:

[TCGA-A2-A0YE-01Z-00-DX1](http://candygram.neurology.emory.edu:8080/histomicstk#?image=5d586d57bd4404c6b1f28640&bounds=53566%2C33193%2C68926%2C40593%2C0)

The user uses a csv file like the one in 
``` plugin_tests/test_files/sample_GTcodes.csv ```
to control pixel values assigned to mask, overlay order of various annotation groups, which groups are considered to be ROIs, etc.

**New requirements:** 
```girder_client``` , ```imageio```, and ```shapely[vectorized]``` are needed as new requirements and have been added to the requirements file.

**Why is this needed?**
There is no available functionality in existing HistomicsTK that does this. The code to perform these tasks is actually non-trivial (albeit simple at heart) and it makes sense to wrap this functionality for users into high-level functions. The code added here handles the following complex situations:
  - Multiple ROIs per item
  - Rotated rectangular annotations and ROIs
  - Polygonal ROIs
  - Overlapping annotations

**Where to look?**

```
|_ histomicstk/
|   |_annotations_and_masks/
|      |_annotation_and_mask_utils.py 
|      |_annotations_to_masks_handler.py
|      |_tests/
|         |_ annotation_and_mask_utils_test.py
|         |_ annotations_to_masks_handler_test.py
|         |_test_files/
|            |_sample_GTcodes.csv
|_ docs/
    |_examples/
       |_annotations_to_masks_handler.ipynb
```

### CODE REVIEW

**Comment by kheffah** [comment]
Annotations to masks successfully passed travis-ci integration. Next steps:
* Migrate to test on a public collection on the candygram server to test edge cases
* Jupyter Notebook for demo usecase
* Add the masks-to-annotations component.

**Comment by kheffah** [comment]
Passes python 3 but not python 2 for some reason.

**Comment by kheffah** [comment]
Passes travis integration!!!

**Comment by kheffah** [comment]
@manthey @dgutman @cooperlab I fixed all issued and added documentation, including a jupyter notebook under docs/examples/ and added all files into the rst documentation. The file is ready to merge whenever you're ready. I'll go ahead and create another branch for the opposite functionality (masks to polygons) in the mean time.

File: ninja [hunk 1]
```diff
@@ -0,0 +1 @@
+/usr/bin/ninja
```
Comment (manthey) [inline ninja]: Did you mean to check this in?

Comment (kheffah) [reply ninja]: @manthey opps, no it must've been added automatically and escaped my attention. I removed it now.

File: plugin_tests/annotation_and_mask_utils_test.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
+# -*- coding: utf-8 -*-
```
Comment (manthey) [inline plugin_tests/annotation_and_mask_utils_test.py]: These new test files are being run as part of the CI ctest process.  Because this repo isn't quite a simple Girder 2.x plugin, tests have to be added explicitly in the `plugin.cmake` file.  Specifically, add appropriate lines like:
```
add_histomicstk_python_test(annotation_and_mask_utils)

add_histomicstk_python_test(annotations_to_masks_handler)
```
I would add this just above the `# front-end tests` section in `plugin.cmake`.

Comment (kheffah) [reply plugin_tests/annotation_and_mask_utils_test.py]: @manthey After these new changes, integration fails for python 2.7 for some weird reason. It says:
```
35: AttributeError: 'module' object has no attribute 'annotations_to_masks_handler_test'
35: Coverage.py warning: No data was collected. (no-data-collected)
23/39 Test #35: server_HistomicsTK.annotations_to_masks_handler ......................***Failed    0.40 sec
test 36
      Start 36: web_client_HistomicsTK.annotations
```
Any idea what's going on?

Comment (manthey) [reply plugin_tests/annotation_and_mask_utils_test.py]: I'm not sure yet -- I'm looking into it.

File: histomicstk/annotations_and_masks/annotations_to_masks_handler.py [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+    """Parse annotations and saves ground truth masks for ALL ROIs.
+
+    Get all ROIs in a single slide. This is a wrapper around get_roi_mask()
+    which should be referred to for implementation details.
+
+    Parameters
+    -----------
+    gc : object
+        girder client object to make requests, for example:
+        gc = girder_client.GirderClient(apiUrl = APIURL)
```
Comment (manthey) [inline histomicstk/annotations_and_masks/annotations_to_masks_handler.py]: In our other tests, we add files that are to big to include in the repo as a sha512 hash that gets downloaded from data.kitware.com (listed in plugin_tests/data).  `plugin.cmake` lists this as external data, and then the test "upload" the file to a test instance of a girder server.  This provides a bit more isolation in the tests (and reduces network traffic, as the files aren't redownloaded if they have already been obtained).

I can refactor this to use that method, if you like.

Comment (kheffah) [reply histomicstk/annotations_and_masks/annotations_to_masks_handler.py]: @manthey Yes please, that would be great!! Thank you!

File: plugin_tests/annotation_and_mask_utils_test.py (hunk 2 above)
Comment (manthey) [reply plugin_tests/annotation_and_mask_utils_test.py]: `FileExistsError` in `annotations_to_masks_handler_test` is a Python3 exception, so it doesn't work in Python 2.  I'll fix it along with the other change I'm making.

Comment (kheffah) [reply plugin_tests/annotation_and_mask_utils_test.py]: Ah I see. thank you :)

**Comment by kheffah** [comment]
# Masks to annotations workflow

**Overview:**

This includes two key functionalities:

- Parsing annotatoins from a groung truth mask into contours (pandas dataframe)

- Parsing contours into a format that is compatible with large_image annotation schema and
ready to push for visualization in Digital Slide Archive HistomicsTK. 


This extends on some of the workflows described in Amgad et al, 2019:

__Mohamed Amgad, Habiba Elfandy, Hagar Hussein, ..., Jonathan Beezley, Deepak R Chittajallu, David Manthey, David A Gutman, Lee A D Cooper, Structured crowdsourcing enables convolutional segmentation of histology images, Bioinformatics, 2019, btz083__


This slide used as a test example:

[TCGA-A2-A0YE-01Z-00-DX1](http://candygram.neurology.emory.edu:8080/histomicstk#?image=5d586d76bd4404c6b1f286ae&bounds=54743%2C32609%2C68828%2C39395%2C0 )

**Where to look?**

```
|_ histomicstk/
|   |
|   |_annotations_and_masks/
|      |_masks_to_annotations_handler.py 
|
|_ plugin_tests/
    |
    |_annotations_and_masks/
    |  |_ masks_to_annotations_handler_test.py
    |
    |_test_files/
       |_sample_GTcodes.csv 
       |_TCGA-A2-A0YE-01Z-00-DX1.8A2E3094-5755-42BC-969D-7F0A2ECA0F39_left-59206_top-33505_mag-BASE.png
       |_sample_contours_df.tsv 
```

@manthey @cooperlab @dgutman I added the opposite functionality (masks to annotations) and merged it with this branch.

**Comment by kheffah** [comment]
@manthey I was about o make a separate PR for this masks to annotations workflow to master in HistomicsTK. The problem is that this would have included all of the changes from this PR as well so would've make it harder for your to check. So for similicity and to make your life easier, I just merged the opposite functionality here.

**Comment by manthey** [comment]
@kheffah Thanks.

**Comment by kheffah** [comment]
@manthey OK I think everythign is set. I checked everythign locally and it works, including PEP8 compliance, pydocstyle, and sphinx build. I also merged your new PR from master. Travis CI stops very early on so it looks like the merged change in master still didn't fix the issue. Other than that, I think everything else is set.

**Comment by kheffah** [comment]
@manthey I'm assuming ``.girder_client_common`` should be part of this repo but was not committed for some reason?

https://github.com/CancerDataScience/HistomicsTK/blob/polygonWorkflows/plugin_tests/annotations_to_masks_handler_test.py#L18

File: plugin_tests/annotation_and_mask_utils_test.py [hunk 4]
```diff
@@ -10,7 +10,7 @@
 from histomicstk.annotations_and_masks.annotation_and_mask_utils import (
     get_image_from_htk_response, get_bboxes_from_slide_annotations)
 
-from .girder_client_common import GirderClientTestCase
+from girder_client_common import GirderClientTestCase
```
Comment (manthey) [inline plugin_tests/annotation_and_mask_utils_test.py]: Python 3 relative imports require the dot.

**Comment by manthey** [comment]
Thanks @kheffah.
